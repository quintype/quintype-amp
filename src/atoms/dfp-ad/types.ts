import { ReactElement } from "react";
import { Story } from "../../types/story";
import { Config } from "../../types/config";

export interface DfpAdTypes {
  width: number;
  height: number;
  "data-slot": string;
  story: Story;
  config: Config;
  children?: JSX.Element[] | JSX.Element | ReactElement | HTMLElement;
  inlineStyles?: object;
}

export interface StyledDfpAdTypes {
  style?: object;
}
