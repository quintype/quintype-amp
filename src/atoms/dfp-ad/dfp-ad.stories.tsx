import * as React from "react";
import { storiesOf } from "@storybook/react";
import { DfpAd } from "./dfp-ad";
import { config, textStory } from "../../__fixtures__";
import { Layout } from "../layout";

const genericAd = {
  width: 300,
  height: 250,
  "data-slot": "/35096353/amptesting/formats/sticky"
};

const invalidAd = {
  width: 300,
  height: 250,
  "data-slot": "/4119129/doesnt-exist"
};

storiesOf("Ad", module)
  .addDecorator((story) => (
    <Layout story={textStory} config={config}>
      {story()}
    </Layout>
  ))
  .add("Simple Ad", () => <DfpAd {...genericAd} />)
  .add("Ad with Placeholder", () => (
    <DfpAd {...genericAd}>
      <div placeholder="true">Loading ...</div>
    </DfpAd>
  ))
  .add("Ad with fallback", () => (
    <DfpAd {...invalidAd}>
      <div fallback="true">Ad failed to load</div>
    </DfpAd>
  ));
