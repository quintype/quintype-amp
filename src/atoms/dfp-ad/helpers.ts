export const getTargetingInfo = ({ story, config }) => {
  const storySections = story.sections || [];
  const sections = storySections.map((section) => section.slug);
  const env = config.publisherConfig.env;
  const publisherName = config.publisherConfig["publisher-name"];
  const targetingObj = {
    targeting: {
      environment: [env],
      "publisher-name": [publisherName],
      sections
    }
  };
  return JSON.stringify(targetingObj);
};
