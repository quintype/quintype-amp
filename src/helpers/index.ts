import ampifyStory from "./ampify-story";
import renderToString from "./render-to-string";
import { genStyles } from "./gen-styles";
import { matchStoryElement } from "./match-story-element";
import { focusedImagePath } from "./image-helpers";

export { renderToString, ampifyStory, focusedImagePath, matchStoryElement, genStyles };
