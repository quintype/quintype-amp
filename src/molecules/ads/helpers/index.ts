import get from "lodash.get";

export const getAdPropsFromConfig = ({ config, adName }) => get(config, ["ampConfig", "doubleclick", adName], null);
export const isEmpty = (obj: object) => Object.keys(obj).length === 0;
export const getDataSlotFromUnitPath = (adProps) => {
  const unitPath = get(adProps, "unit-path", null);
  if (unitPath) {
    delete adProps["unit-path"];
    adProps["data-slot"] = unitPath;
  }
  return adProps;
};

// export const getAdPropsFromConfig = ({ config, adName }) => get(config, ["ampConfig", "ads", adName], null);
//
// config = {  // in the future, this is how ads will be in the config
//   ampConfig: {
//     ...
//     ads: {
//       "top-ad": {
//         type: "industrybrains",
//         width: 300,
//         height: 250,
//         "data-width": "300",
//         "data-height": "250",
//         "data-cid": "19626-3798936394"
//       },
//       "bottom-ad": {
//         width: "320",
//         height: "50",
//         type: "doubleclick",
//         "data-slot": "/35096353/amptesting/formats/sticky"
//       }
//     },
//     ...
//   }
// }
