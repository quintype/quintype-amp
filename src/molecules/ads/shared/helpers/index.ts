import get from "lodash.get";
import cloneDeep from "lodash.clonedeep";

const isEmpty = (obj: object) => Object.keys(obj).length === 0;

export const getPropsForDfpAd = ({ overridingProps, config, adName }) => {
  if (isEmpty(overridingProps)) {
    const clone = cloneDeep(config);
    const adPropsFromConfig = get(clone, ["ampConfig", "doubleclick", adName], null);
    const unitPath = get(adPropsFromConfig, "unit-path", null);

    if (!adPropsFromConfig) throw new Error(`No props defined for ${adName} in config`);
    if (!unitPath) throw new Error(`unit-path not defined for ad ${adName} in config`);
    adPropsFromConfig["data-slot"] = unitPath; // renames unit-path to data-slot
    delete adPropsFromConfig["unit-path"];
    return adPropsFromConfig;
  } else return overridingProps;
};
