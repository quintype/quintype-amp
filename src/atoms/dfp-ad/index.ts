import { DfpAd } from "./dfp-ad";
export { DfpAd };
