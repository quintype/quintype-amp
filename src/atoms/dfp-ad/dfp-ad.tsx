import React, { Fragment } from "react";
import { Helmet } from "react-helmet";
import { DfpAdTypes } from "./types";
import { withStoryAndConfig } from "../../context";
import { getTargetingInfo } from "./helpers";

const DfpAdBase = ({ story, config, children, prefetchScript, ...rest }: DfpAdTypes) => {
  return (
    <Fragment>
      <Helmet>
        {prefetchScript ? (
          <script
            rel="preload"
            crossorigin="anonymous"
            async={undefined}
            custom-element="amp-ad"
            src="https://cdn.ampproject.org/v0/amp-ad-0.1.js"
          />
        ) : (
          <script async={undefined} custom-element="amp-ad" src="https://cdn.ampproject.org/v0/amp-ad-0.1.js" />
        )}
      </Helmet>
      <amp-ad type="doubleclick" json={getTargetingInfo({ story, config })} {...rest}>
        {children}
      </amp-ad>
    </Fragment>
  );
};

export const DfpAd = withStoryAndConfig(DfpAdBase);
