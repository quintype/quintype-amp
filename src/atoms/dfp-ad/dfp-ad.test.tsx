/**
 * @jest-environment node
 */

// FYI mount wont work here as this runs in node env
import { getTargetingInfo } from "./helpers";
import { shallow } from "enzyme";
import { DfpAd } from "./dfp-ad";
import React from "react";

describe("dfp-ad component", () => {
  it("Should match snapshot", () => {
    const wrapper = shallow(<DfpAd />);
    expect(wrapper).toMatchSnapshot();
  });
  it("should render dfp with custom styles", () => {
    const wrapper = shallow(<DfpAd inlineStyles={{ border: "2px solid red" }} />);
    expect(wrapper.find("div").prop("style")).toStrictEqual({ border: "2px solid red" });
  });
});
test("getTargetingInfo helper", async () => {
  const dummyStory = {
    sections: [{ slug: "news" }, { slug: "politics" }]
  };
  const dummyConfig = {
    publisherConfig: {
      env: "staging",
      "publisher-name": "Arnab Goswami's Republic TV"
    }
  };
  const targetingJson = getTargetingInfo({ story: dummyStory, config: dummyConfig });
  const obj = {
    targeting: {
      environment: ["staging"],
      "publisher-name": ["Arnab Goswami's Republic TV"],
      sections: ["news", "politics"]
    }
  };
  expect(JSON.parse(targetingJson)).toMatchObject(obj);
});
