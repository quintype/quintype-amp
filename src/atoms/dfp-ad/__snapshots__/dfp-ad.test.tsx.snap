// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dfp-ad component Should match snapshot 1`] = `
<Component
  config={Object {}}
  story={Object {}}
/>
`;
